### PR TITLE
Populate dash

### PR DESCRIPTION
**Pull request by DomLaw82**
**SQLConnection.py**
- Add threading with lock object applied to the query and execute functions to resolve "Recursive use of cursors not allowed" error when trying to use the SQLConnection class to update each of the dashboard sections in real time
- Re-adding the execute function to collect single value to update values displayed for dashboard metrics, (query to update graph, execute to update raw values displayed)

**app.py**
- Change port running dash server to 8080

**style.css**
- Colouring webpage
- Hover animations for the page switch button
- Emojis in the metric descriptions

**callbacks.py**
- Created callbacks to update the graphs on each page live, once a second
- Callbacks to update hr, rpm, resistance and duration values in real time

**index.py**
- Add header element to page
- Contains interval element for updates and url component for page switching

**current_view.py/recent_view.py**
- Layout changes
- Adding emojis to metric descriptions

**visualistions.py**
- Adding colouring to the grid, plot background and tick labels

### CODE REVIEW

File: dash/SQLConnection.py [hunk 1]
```diff
@@ -1,11 +1,14 @@
 import sqlite3
 import pandas as pd
+import threading
+
+lock = threading.Lock()
```
Comment (sebjshaw) [inline dash/SQLConnection.py:5]: this is neat, didn't know about this. Have you added `threading` to the `requirements.txt`?

File: dash/assets/style.css [hunk 2]
```diff
@@ -1,14 +1,26 @@
...
 	width: 8vw;
+	background-color: #368f8b;
+	color: #f3dfc1;
+}
+#view_switch:hover {
+	border-color: #8d5b4c 4px solid;
+	background-color: #23625f;
+	width: 8.2vw;
+	height: 8.2vh;
+	transition: 0.1s;
```
Comment (sebjshaw) [inline dash/assets/style.css:23]: I assume this makes the view look cool when you hover over it? Nice!

File: dash/SQLConnection.py [hunk 3]
```diff
@@ -19,9 +22,34 @@ def query(self, query:str) -> pd.DataFrame:
...
+			lock.acquire(True)
+			res = self.curs.execute(query)
+			res = res.fetchall()
+			print(f"'{query}' executed")
+			return res
+		except Exception as e:
+			print(e)
+			print('\n')
+		finally:
+			lock.release()
```
Comment (sebjshaw) [inline dash/SQLConnection.py:55]: `query` and `execute` both necessary but maybe change their names to be more descriptive? `query` returns a df and `execute` returns a list so maybe call them `q_df` and `q_list` or something like that

File: dash/callbacks.py [hunk 4]
```diff
@@ -44,89 +44,148 @@ def change_link(n):
...
+	),
+	[
+		Input(
+			'interval_component', 'n_intervals'
+		)
+	]
+)
+def update_resistance(n):
+	duration = int(sql.execute("""SELECT duration from current_ride""")[-1][0])
+	return str(timedelta(seconds=duration))[2:]
```
Comment (sebjshaw) [inline dash/callbacks.py:191]: Call backs look good!

File: dash/views/current_view.py [hunk 5]
```diff
@@ -7,27 +7,55 @@
...
+									children = [
+										html.H2('DURATION ⏱️',id='time_text', className='bit_title'),
+										html.H2(id='time_value')
+									],
+									id='time',
+									className='bit'
+								),
+								html.Div(
+									children = [
+										html.H2('RESISTANCE 😤',id='resistance_text', className='bit_title'),
```
Comment (sebjshaw) [inline dash/views/current_view.py:27]: Nice emojis!

File: dash/views/current_view.py [hunk 6]
```diff
@@ -7,27 +7,55 @@
...
+									children = [
+										html.H2('RPM 🔄',id='rpm_text', className='bit_title'),
+										html.H2(id='rpm_value'),
+									],
+									id='rpm',
+									className='bit'
+								),
+								html.Div(
+									children = [
+										html.H2('HEART RATE 🫀',id='heart_rate_text', className='bit_title'),
```
Comment (sebjshaw) [inline dash/views/current_view.py:48]: Maybe change to a more cartoon heart like ❤️? Personal preference at the end of the day so up to you!

**Review by sebjshaw — APPROVED** [review]

File: dash/SQLConnection.py (hunk 1 above)
Comment (DomLaw82) [reply dash/SQLConnection.py:5]: threading is a python module, so it doesn't need to be in there, the requirements.txt just installs 3rd party modules

File: dash/SQLConnection.py (hunk 3 above)
Comment (DomLaw82) [reply dash/SQLConnection.py:55]: I like that, distinguish their difference in their names

Comment (DomLaw82) [reply dash/SQLConnection.py:55]: changed to get_df and get_list

**Pull request merged**